### DISCUSSION TITLE
Correct way to handle "AccessDenied" and other errors

### DISCUSSION THREAD

**Discussion by echelonh** [awslabs/aws-sdk-rust#1211] [Q&A] [answered]
Hi there!

I'm trying to write code that assumes a role and then goes on to perform various actions with that role (e.g. read from s3 bucket).

It's pretty similar to the examples shown here:
https://github.com/awslabs/aws-sdk-rust/tree/main/examples/examples/sts/src/bin

Since the user can provide a role that I don't have permissions to assume - I want to catch an "AccessDenied" error and handle it - print something informative to the user about not having permissions.

Unfortunately, after looking at most examples online - I didn't see any way to properly handle it and access the error's code easily. Not by attempting to use `provider.provide_credentials()` which yiels `CredentialsError`, nor by calling `get_caller_identity()` which yields `DispatchFailure`. 

The errors are a huge nested mess, and I honestly didn't understand how to get the access code - ugly code or not.

I'd appreciate any help - thanks a lot! :pray:

**Comment by ysaito1001** [comment, marked as answer]
While this isn't strictly a type-based approach, one way to determine if the returned service error is related to `AccessDenied` is by examining the [DisplayErrorContext](https://docs.aws.amazon.com/sdk-for-rust/latest/dg/error-handling.html#displayErrorContext). [These](https://github.com/search?q=repo%3Aawslabs%2Faws-sdk-rust+DisplayErrorContext+path%3A%2F%5Eexamples%5C%2Fexamples%5C%2F%2F&type=code) are examples in our repository showing how you can use it. You should then be able to search for an error string like `AccessDenied`.

**Reply by echelonh** [reply]
Hey,

Thanks for the suggestion, but this feels very far-off from what I'm trying to do. I could "grep" for certain strings, but this is a very dangerous solution, and it can be quite inaccurate.

I'm looking for something that's even slightly more formal. Also, it's very possible that `AccessDenied` is not the only code I'm after.

Thanks!

**Reply by ysaito1001** [reply]
> handle it

You can also try using [`.into_service_error`](https://docs.aws.amazon.com/sdk-for-rust/latest/dg/error-handling.html#serviceErrors) to turn it into a `ServiceError` which you can then match on.

> print something informative to the user about not having permissions

For printing purposes, you should just be able to use `DisplayErrorContext` to print the underlying error message.

**Reply by echelonh** [reply]
Thanks for the suggestions! :)
I really appreciate you taking the time to help :pray: 

I've already tried using `.into_service_error()` before writing this post, but after lots of trial and error where I didn't get the correct code, I figured this is the error hierarchy I get (in this example, for `provide_credentials`):
```
ProviderError(ProviderError { 
  source: ServiceError(ServiceError {
    source: Unhandled(Unhandled {
      source: ErrorMetadata {
```
Only `Unhandled` and `ErrorMetadata` has the `code` field properly set - not sure what the next step after `.into_service_error` should be.

Regarding the printing purposes - The consumer of the data I'm sending shouldn't be aware of anything too specific in AWS, definitely not in such detail.
`DisplayErrorContext` is good for my own debugging purposes, but can't be used as something informative or a protocol for my internal services to rely on.

**Comment by stijnherfst** [comment]
Something like this should work pretty well:
```rs
fn is_aws_access_denied<E: std::error::Error + ProvideErrorMetadata>(err: &E) -> bool {
    if let Some(code) = err.code() {
        return code.contains("AccessDenied") || code.contains("UnauthorizedOperation");
    }

    let mut source = err.source();
    while let Some(e) = source {
        if let Some(meta) = e.downcast_ref::<aws_smithy_types::error::ErrorMetadata>()
            && let Some(code) = meta.code()
        {
            return code.contains("AccessDenied") || code.contains("UnauthorizedOperation");
        }
        source = e.source();
    }
    false
}
```
